cpp/ConnectionPool.cpp

```diff
@@ -0,0 +1,330 @@
+#include "ConnectionPool.h"
+#include "fileUtils.h"
+#include "sqlite3.h"
+#include "sqliteBridge.h"
+#include "sqliteExecute.h"
+#include <fstream>
+#include <iostream>
+
+ConnectionPool::ConnectionPool(std::string dbName, std::string docPath,
+                               unsigned int numReadConnections)
+    : dbName(dbName), maxReads(numReadConnections) {
+
+  onContextCallback = nullptr;
+  isConcurrencyEnabled = maxReads > 0;
+
+  struct ConnectionState writeCon;
+  writeCon.connection = nullptr;
+  writeCon.currentLockId = EMPTY_LOCK_ID;
+
+  writeConnection = writeCon;
+
+  // Open the write connection
+  genericSqliteOpenDb(dbName, docPath, &writeConnection.connection,
+                      SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE |
+                          SQLITE_OPEN_NOMUTEX);
```

Comment:
Since each connection is used from different threads (even if it's not at the same time in most cases), `SQLITE_OPEN_FULLMUTEX` would be safer.

Reply:
The connections now open with `SQLITE_OPEN_FULLMUTEX` all the tests still pass correctly. I'm not actually sure why it was using `SQLITE_OPEN_NOMUTEX`.

This module did originally open connections with `SQLITE_OPEN_FULLMUTEX` and it's asynchronous nature allowed different threads to access the connections (with no locks for preventing simultaneous asynchronous operations. I assume if a conflict occurred some operations would fail and throw and error).